### PR TITLE
Bug 1876855: Document storage exceptions to etcd restore

### PR DESCRIPTION
**Pull request by gnufied**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1876855

### CODE REVIEW

File: modules/dr-restoring-cluster-state.adoc [hunk 1]
```diff
@@ -304,3 +304,15 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+Some examples are:
+
+1. Mysql database was running in a Pod backed up by a PV object. But after the snapshot was taken - the Mysql Pod and corresponding PV were deleted and hence underlying volume on storage provider would be deleted. Restoring Openshift from snapshot is not going to bring back the volume on storage provider and hence is not going to result in running Mysql Pod(even though Pod will try to start but will keep failing) and a manual action must be taken to restore mysql Pod. In this case manual action could be restoring the volume on storage provider via volume snapshots and then editing PV object to point to new volume.
+
+2. You have a pod P1 running using volume A which is attached to node X. If the snapshot is taken and another pod uses the same volume on node Y and a etcd restore is performed then pod P1 may not able to start correctly because volume is still attached to node Y but {product-title} does know about it and hence is not going to detach it. In this case volume must be manually detached from node Y, so as volume can attach on node X and pod P1 can start.
```
Comment (smarterclayton) [inline modules/dr-restoring-cluster-state.adoc:318]: Wont the storage provider eventually reconcile?  Give an example of a storage provider this won’t work on?

Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc:318]: It won't work for CSI(all CSI volumes) or GCE-PD intree driver for example. For intree driver's volume's attachment status is recorded in `node.Status` and similarly for CSI volumes in separate `VolumeAttachment` object. If a etcd restore causes `VolumeAttachment` to be removed or `node.Status` to be wiped - then attach/detach controller will not try and detach the volumes.   We have partially fixed this for some intree drivers by implementing a dangling volume mechanism but it isn't completely solved for CSI volumes - https://github.com/kubernetes/kubernetes/issues/80488 and some intree drivers (like GCE-PD) don't implement dangling volume mechanism at all. 

There is a opposite of this - which is, if when the snapshot was taken volume was attached to node X but since then pod got deleted and volume was detached but then snapshot is restored and pod starts on node X but since attach-detach controller thinks volume is *actually* attached already (because `node.Status` tells it so), it won't try to attach the volume again. This is also partially solved for some volume types but for CSI volumes `LIST_VOLUME` is an optional capability.

Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc:318]: Please note even though - we *now* have a dangling volume mechanism for CSI, if corresponding VolumeAttachment objects does not exist after restore from snapshot, volume will not be detached to correct the attachment issue.

**Comment by jsafrane** [comment]
In general I like the approach. Please explicitly mention "ectd snapshot" and "volume snapshot" everywhere, as it's heavily overused term and it must be crystal clear which snapshot is which.

File: modules/dr-restoring-cluster-state.adoc [hunk 2]
```diff
@@ -304,3 +304,19 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+3. If cloudprovider or storage provider credentials were updated since snapshot was taken, any CSI drivers or operators that depend on the those credentials will not work and hence you may have to manually update credentials required by those drivers or operators.
+
```
Comment (jsafrane) [inline modules/dr-restoring-cluster-state.adoc:321]: I think it would make sense to be really explicit and have an item that content of PVs will be different than when the etcd snapshot was taken.

File: modules/dr-restoring-cluster-state.adoc [hunk 3]
```diff
@@ -304,3 +304,19 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+4. Local Storage Operator creates symlinks for each persistent volume it manages from `/dev/disk/by-id` or `/dev` directories. If a device is removed or renamed from {product-title} nodes since snapshot was taken, the local volume Persistent Volumes(PVs) may refer to devices which no longer exist. To workaround this problem - admin must manually remove the PVs with invalid devices and remove symlink from respective nodes and delete `LocalVolume` or `LocalVolumeSet` objects. See section on deleting Local Storage Operator Resources in {product-title} documentation.
```
Comment (jsafrane) [inline modules/dr-restoring-cluster-state.adoc:322]: Maybe a note that especially for local volumes it may happen that a StatefulSet S1 with PVC1 uses PV1 and stores something not sensitive there (funny cat videos??) when the etcd snapshot is taken. Later on, the whole S1 is decommissioned and the disk used for PV1 is later used for sensitive database, such as database of customers. When the etcd snapshot is restored, S1 + PVC1 + PV1 is restored, having access to the sensitive data instead of cat videos.

Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc:322]: Good point. Though this is not common to local storage right? If you have a PV that you are using to store database files and you restore cluster to a version where it was used for storing images (and perhaps has public http access) - then suddenly you published your database to the whole world.

File: modules/dr-restoring-cluster-state.adoc (hunk 2 above)
Comment (gnufied) [reply modules/dr-restoring-cluster-state.adoc:321]: done.

File: modules/dr-restoring-cluster-state.adoc [hunk 4]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
 ----
 
 Note that it might take several minutes after completing this procedure for all services to be restored. For example, authentication by using `oc login` might not immediately work until the OAuth server pods are restarted.
+
+[id="dr-scenario-2-restoring-cluster-state_storage_{context}"]
+= Issues and workarounds in restoring persistent storage state
+
+If your {product-title} uses persistent storage of any form, typically a state of the world is stored outside etcd. It could be a ElasticSearch
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:311]: Is it just one state of the world that's stored? If so, maybe `...typically the state of the world`

A nitpick, but `a ElasticSearch` -> `an ElasticSearch`.

File: modules/dr-restoring-cluster-state.adoc [hunk 5]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+cluster running in a Pod or a database running in a statefulset. When {product-title} is restored from etcd backup, we are also restoring status of those workloads in Openshift. But depending on how old the etcd snapshot is - that status may be invalid or outdated.
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:312]: "restored from the etcd backup"

I'd avoid starting a sentence with `But`. You could drop this word and have the sentence read - "The status may be invalid or outdated depending on how old the etcd snapshot is."

File: modules/dr-restoring-cluster-state.adoc [hunk 6]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+Some examples are:
+
+1. Mysql database was running in a Pod backed up by a PV object. But after the etcd snapshot was taken - the Mysql Pod and corresponding PV were deleted and hence underlying volume on storage provider would be deleted. Restoring Openshift from etcd snapshot is not going to bring back the volume on storage provider and hence is not going to result in running Mysql Pod(even though Pod will try to start but will keep failing) and a manual action must be taken to restore mysql Pod. In this case manual action could be restoring the volume on storage provider via volume snapshots and then editing PV object to point to new volume.
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:316]: I'd avoid starting the sentence with `But`. This sentence could read - "After the etcd snapshot was taken the Mysql Pod and corresponding PV were deleted, and the underlying volume on the storage provider was also deleted."

Personal nitpick, but I'd reword the rest of this bit to say - "Restoring OpenShift from an etcd snapshot is not going to bring back the volume on the storage provider, and will not result in a running Mysql Pod, despite the Pod repeatedly attempting to start. A manual action must be taken to restore this Pod, such as restoring the volume on the storage provider and then editing the PV to point to the new volume."

File: modules/dr-restoring-cluster-state.adoc [hunk 7]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+2. You have a pod P1 running using volume A which is attached to node X. If the etcd snapshot is taken and another pod uses the same volume on node Y and a etcd restore is performed then pod P1 may not able to start correctly because volume is still attached to node Y but {product-title} does know about it and hence is not going to detach it. In this case volume must be manually detached from node Y, so as volume can attach on node X and pod P1 can start.
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:318]: The second and third sentences are a bit clunky. I'd suggest rewording it to be - "If the etcd snapshot is taken while another pod uses the same volume on node Y, then when the etcd restore is performed pod P1 may not be able to start correctly due to the volume still being attached to node Y. {product-title} is aware of the attachment, and will not automatically detach it." 

Last sentence -"...case the volume".

File: modules/dr-restoring-cluster-state.adoc [hunk 8]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+3. If cloudprovider or storage provider credentials were updated since etcd snapshot was taken, any CSI drivers or operators that depend on the those credentials will not work and hence you may have to manually update credentials required by those drivers or operators.
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:320]: I'd break this into two sentences - "...those credentials will not work. You must manually update the credentials required by these drivers or operators."

File: modules/dr-restoring-cluster-state.adoc [hunk 9]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+4. Local Storage Operator creates symlinks for each persistent volume it manages from `/dev/disk/by-id` or `/dev` directories. If a device is removed or renamed from {product-title} nodes since etcd snapshot was taken, the local volume Persistent Volumes(PVs) may refer to devices which no longer exist. To workaround this problem - admin must manually remove the PVs with invalid devices and remove symlink from respective nodes and delete `LocalVolume` or `LocalVolumeSet` objects. See section on deleting Local Storage Operator Resources in {product-title} documentation.
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:322]: We used PV back in point 1. I'd move the acronym definition up there.

It may be best to break apart the actions that an admin has to do into an unordered list beneath this point, as we're defining three things in a single sentence.

File: modules/dr-restoring-cluster-state.adoc [hunk 10]
```diff
@@ -304,3 +304,21 @@ etcd-ip-10-0-173-171.ec2.internal                2/2     Running     0
...
+
+5. Contents of Persistent Volumes(PVs) is never part of etcd snapshots and hence when restoring {product-title} cluster from etcd snapshot could result in non-critical workloads to gain access to critical data (or vice-versa).
```
Comment (huffmanca) [inline modules/dr-restoring-cluster-state.adoc:324]: We should be defining PV above this point, so I think it's safe to remove the definition here.

I'd also suggest breaking this into multiple sentences - "The contents of PVs are never part of the etcd snapshot. When restoring an {product-title} cluster from the etcd snapshot it could result in non-critical workloads gaining access to critical data (or vice-versa)."

**Review by huffmanca — COMMENTED** [review]
The content is good, but I have some nitpicks and wording suggestions for readability. Thank you!

**Comment by jsafrane** [comment]
From my POV, all the issues we want to emphasize are in this PR. I let others comment on the style.

File: modules/dr-restoring-cluster-state.adoc (hunk 5 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:312]: - "Pod" should be lowercase ("pod").
- Change the sentence "When OCP is restored..." to "When restoring from an etcd backup, the status of the workloads in OCP is also restored."
- Change the sentence "But depending..." to "However, if the etcd snapshot is old, the status may be invalid or outdated."

File: modules/dr-restoring-cluster-state.adoc (hunk 6 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:316]: What are the example of? Stale ectd backups? If so, provide more clarification in the lead-in sentence "Some examples are:"

Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:316]: Introduce the term "PV" and lowercase for "pod". So..."Mysql database was running in a pod backed up by a persistent volume (PV) object."

Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:316]: I like a lot of Christian's rewrite above, but would recommend a few changes to:
"Restoring OpenShift from an etcd snapshot does not bring back the volume on the storage provider, and does not produce a running Mysql pod, despite the pod repeatedly attempting to start. You must manually restore this pod by restoring the volume on the storage provider, and then editing the PV to point to the new volume."

File: modules/dr-restoring-cluster-state.adoc (hunk 7 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:318]: Working with Christian's rewrite:
"Pod P1 is running using volume A, which is attached to node X. If the etcd snapshot is taken while another pod is using the same volume on node Y, then when the etcd restore is performed pod P1, it may not be able to start correctly due to the volume still being attached to node Y. {product-title} is aware of the attachment, and does not automatically detach it. When this occurs, the volume must be manually detached from node Y, so the volume can attach on node X, and then pod P1 can start."

File: modules/dr-restoring-cluster-state.adoc (hunk 9 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:322]: "Local Storage Operator creates symlinks for each persistent volume it manages from `/dev/disk/by-id` or `/dev` directories. If a device is removed or renamed from {product-title} nodes after the etcd snapshot is taken, the local PVs may refer to devices that no longer exist. To fix this problem, an admin must:
1. Manually remove the PVs with invalid devices.
2. Remove symlink from respective nodes.
3. Delete `LocalVolume` or `LocalVolumeSet` objects (see Storage -> Configuring persistent storage -> Persistent storage using local volumes
 -> Deleting the Local Storage Operator Resources).

File: modules/dr-restoring-cluster-state.adoc (hunk 10 above)
Comment (lpettyjo) [reply modules/dr-restoring-cluster-state.adoc:324]: I'd change to: "The contents of PVs are never part of the etcd snapshot. When restoring an {product-title} cluster from the etcd snapshot, non-critical workloads could gain access to critical data (or vice-versa)."

However, point 5 seems to be different from the preceding 4 points. It's more of warning than about a scenario that creates a stale/wrong etcd. I'd pull this out of the list and make it an "Important" admonition.
